pkg/transportrequest/solman/upload.go

```diff
@@ -0,0 +1,128 @@
+package solman
+
+import (
+	"fmt"
+	"github.com/SAP/jenkins-library/pkg/config/validation"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/pkg/errors"
+	"strings"
+)
+
+// FileSystem interface collecting everything which is file system
+// related and needed in the context of a SOLMAN upload.
+type FileSystem interface {
+	FileExists(path string) (bool, error)
+}
+
+// UploadAction Collects all the properties we need for the deployment
+type UploadAction struct {
+	Connection         Connection
+	ChangeDocumentID   string
+	TransportRequestID string
+	ApplicationID      string
+	File               string
+	CMOpts             []string
+}
+
+// Action collects everything which is needed to perform a SOLMAN upload
+type Action interface {
+	WithConnection(Connection)
+	WithChangeDocumentID(string)
+	WithTransportRequestID(string)
+	WithApplicationID(string)
+	WithFile(string)
+	WithCMOpts([]string)
+	Perform(fs FileSystem, command Exec) error
+}
+
+// WithConnection specifies all the connection details which
+// are required in order to connect so SOLMAN
+func (a *UploadAction) WithConnection(c Connection) {
+	a.Connection = c
+}
+
+// WithChangeDocumentID specifies the change document which
+// receives the executable.
+func (a *UploadAction) WithChangeDocumentID(id string) {
+	a.ChangeDocumentID = id
+}
+
+// WithTransportRequestID specifies the transport request which
+// receives the executable.
+func (a *UploadAction) WithTransportRequestID(id string) {
+	a.TransportRequestID = id
+}
+
+// WithApplicationID specifies the application ID.
+func (a *UploadAction) WithApplicationID(id string) {
+	a.ApplicationID = id
+}
+
+// WithFile specifies the executable which should be uploaded into
+// a transport on SOLMAN
+func (a *UploadAction) WithFile(f string) {
+	a.File = f
+}
+
+// WithCMOpts sets additional options for calling the
+// cm client tool. E.g. -D options. Useful for troubleshooting
+func (a *UploadAction) WithCMOpts(opts []string) {
+	a.CMOpts = opts
+}
+
+// Perform performs the SOLMAN upload
+func (a *UploadAction) Perform(fs FileSystem, command Exec) error {
+
+	log.Entry().Infof("Deploying artifact '%s' to '%s'.",
+		a.File, a.Connection.Endpoint)
+
+	missingParameters, err := validation.FindEmptyStringsInConfigStruct(*a)
+
+	if err == nil {
+		notInitialized := len(missingParameters) != 0
+		if notInitialized {
+			err = fmt.Errorf("the following parameters are not available %s", missingParameters)
```

Comment:
I doub't that the quotes are helpful here since we have `[ ... ]` which we get since it is a slice. But in general I agree having quotes is better since a troubleshooter can clearly see that there is something "" "missing".